`assertEquals` shows an incorrect diff

**Describe the bug**

The diff shown by `assertEquals` is not correct if the number of newlines differs in the left and right hand side of the assertion, and if there are repeated lines in the left hand side.

**Steps to Reproduce**

```
import { assertEquals } from "https://deno.land/std@0.170.0/testing/asserts.ts";

assertEquals("bar\nHello\nbar\nWorld1", "Hello2\nWorld");
```

results in the following output:

```
error: Uncaught AssertionError: Values are not equal:


    [Diff] Actual / Expected


-   bar\n
-   Hello\n
-   bar\n
-   World1
+   Hello2\n
+   World


  throw new AssertionError(message);
        ^
    at assertEquals (https://deno.land/std@0.170.0/testing/asserts.ts:190:9)
    at file:///mnt/artemis/Projects/github.com/denoland/deno_std/assert.ts:3:1
```


**Expected behavior**

The diff should look like this:

```
error: Uncaught AssertionError: Values are not equal:


    [Diff] Actual / Expected


-   bar\n
-   Hello\n
-   bar\n
-   World1
+   Hello2\n
+   World


  throw new AssertionError(message);
        ^
    at assertEquals (https://deno.land/std@0.170.0/testing/asserts.ts:190:9)
    at file:///mnt/artemis/Projects/github.com/denoland/deno_std/assert.ts:3:1
```
**Environment**

- OS: Ubuntu 20.04
- deno version: 1.29.1
- std version: 0.170.0

The issue seems to have started from #948 (v0.100.0, the output is different from the current version, but still wrong)

v0.99.0 doesn't have this problem (though multiline strings are displayed less nicely)

```js
import { assertEquals } from "https://deno.land/std@0.99.0/testing/asserts.ts";

assertEquals("bar\nHello\nbar\nWorld1", "Hello2\nWorld");
```

```shellsession
$ deno run c.js 
error: Uncaught AssertionError: Values are not equal:


    [Diff] Actual / Expected


-   "bar\nHello\nbar\nWorld1"
+   "Hello2\nWorld"
```

cc @lowlighter

--- a/assert/_diff.ts
+++ b/assert/_diff.ts
@@ -342,18 +342,21 @@ export function diffstr(A: string, B: string) {
   }
 
   // Compute word-diff
-  const aLines = added.length < removed.length ? added : removed;
-  const bLines = aLines === removed ? added : removed;
+  const hasMoreRemovedLines = added.length < removed.length;
+  const aLines = hasMoreRemovedLines ? added : removed;
+  const bLines = hasMoreRemovedLines ? removed : added;
   for (const a of aLines) {
     let tokens = [] as Array<DiffResult<string>>,
       b: undefined | DiffResult<string>;
     // Search another diff line with at least one common token
     while (bLines.length) {
       b = bLines.shift();
-      tokens = diff(
+      const tokenized = [
         tokenize(a.value, { wordDiff: true }),
         tokenize(b?.value ?? "", { wordDiff: true }),
-      );
+      ] as string[][];
+      if (hasMoreRemovedLines) tokenized.reverse();
+      tokens = diff(tokenized[0], tokenized[1]);
       if (
         tokens.some(({ type, value }) =>
           type === DiffType.common && value.trim().length

--- a/assert/_diff_test.ts
+++ b/assert/_diff_test.ts
@@ -314,3 +314,32 @@ Deno.test({
     ]);
   },
 });
+
+Deno.test({
+  name: "multiline diff with more removed lines",
+  fn() {
+    const diffResult = diffstr("a\na", "e");
+    assertEquals(diffResult, [
+      {
+        type: DiffType.removed,
+        value: "a\\n\n",
+      },
+      {
+        type: DiffType.removed,
+        value: "a\n",
+        details: [
+          { type: DiffType.removed, value: "a" },
+          { type: DiffType.common, value: "\n" },
+        ],
+      },
+      {
+        type: DiffType.added,
+        value: "e\n",
+        details: [
+          { type: DiffType.added, value: "e" },
+          { type: DiffType.common, value: "\n" },
+        ],
+      },
+    ]);
+  },
+});

